From 852b7bbda73f72733b26a39413705858c2cfc4f4 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Fri, 17 Jun 2016 16:37:06 -0700
Subject: [PATCH 0968/1566] drm: xilinx: dp: Set maximum swing / preemphsis
 level to 2 for ZynqMP

commit  7b8840fd11500498d3d0eb1203c108469de8331d from
https://github.com/Xilinx/linux-xlnx.git

The DP spec defines that the level 3 is optional to support for both
voltage swing and preemphasis. Thus, set the maximum voltage swing
and preemphasis level to 2 for ZynqMP DP subsystem.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_dp.c |   13 +++++++------
 1 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
index 279e33f..ffdb1a6 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
@@ -454,7 +454,7 @@ static void xilinx_drm_dp_adjust_train(struct xilinx_drm_dp *dp,
 {
 	u8 *train_set = dp->train_set;
 	u8 voltage = 0, preemphasis = 0;
-	u8 max_preemphasis;
+	u8 max_voltage, max_preemphasis;
 	u8 i;
 
 	for (i = 0; i < dp->mode.lane_cnt; i++) {
@@ -468,14 +468,15 @@ static void xilinx_drm_dp_adjust_train(struct xilinx_drm_dp *dp,
 			preemphasis = p;
 	}
 
-	if (voltage >= DP_TRAIN_VOLTAGE_SWING_LEVEL_3)
-		voltage |= DP_TRAIN_MAX_SWING_REACHED;
-
+	max_voltage = (dp->dp_sub) ? DP_TRAIN_VOLTAGE_SWING_LEVEL_2 :
+				     DP_TRAIN_VOLTAGE_SWING_LEVEL_3;
+	if (voltage >= max_voltage)
+		voltage = max_voltage | DP_TRAIN_MAX_SWING_REACHED;
 	max_preemphasis = (dp->dp_sub) ? DP_TRAIN_PRE_EMPH_LEVEL_2 :
 					 DP_TRAIN_PRE_EMPH_LEVEL_3;
-
 	if (preemphasis >= max_preemphasis)
-		preemphasis |= DP_TRAIN_MAX_PRE_EMPHASIS_REACHED;
+		preemphasis = max_preemphasis |
+			      DP_TRAIN_MAX_PRE_EMPHASIS_REACHED;
 
 	for (i = 0; i < dp->mode.lane_cnt; i++)
 		train_set[i] = voltage | preemphasis;
-- 
1.7.5.4

